### ISSUE TITLE
Angular Service Worker should prefetch assets in parallel

### ISSUE THREAD

**Issue by laurentgoudet** [angular/angular#39491]
# 🚀 feature request

### Relevant Package

This feature request is for @angular/service-worker.

### Description

The Angular Service Worker currently prefetches a new version assets one by one, as mentioned in https://github.com/angular/angular/blob/dd0ba3f4abb5cf467e55bf660b70965ba7f2881e/packages/service-worker/worker/src/assets.ts#L510

> // Cache all known resources serially. As this reduce proceeds, each Promise waits
> // on the last before starting the fetch/cache operation for the next request. Any
> // errors cause fall-through to the final Promise which rejects.

Doing so with a medium/large application that heavily leverage code-splitting in inefficient: in our case we have ~340 assets to prefetch, for a total application size of only 3.0MB, yet upgrades a taking several minutes (sometimes above > 10 minutes) to be ready.

This leads to app upgrades being slower to propagate as they should be, as users have to stick around several minutes within the app for us to be able to switch to (or notify them about) the new version.

### Describe the solution you'd like

New versions assets should be prefetched by the Service Worker in parallel. 3MB only takes a second to download on the average broadband connection thanks for h2+ parallelism, and the Angular Service Worker should leverage that.

### Describe alternatives you've considered

I am unused if the current behavior was designed on purpose.

**Comment by gkalpak** [comment]
I assume this was done to prevent network congestion (@alxhub could confirm).
I still believe it should be the default behavior (or maybe have a few parallel requests to speed things up a bit). But maybe we could allow specifying the desired parallelism in the asset-group entry in `ngsw-config.json`.

This is not a high priority, so not sure when/if we are going to get to it, but if anyone is interested in working on PR I would be happy to help/review.

**Comment by dylhunn** [comment]
We consider this degraded performance a bug in service worker. The user should not have to remain on the app for several minutes to get the update.

**Comment by jvitor83** [comment]
This should be a priority. The prefetch should behave like the browser once it hits the page (downloading the assets at the same time), otherwise the loading will be slower than without service worker.

**Comment by gkalpak** [comment]
FWIW, prefetching does not affect the loading time of the page, since it happens in the background (independently of the page loading process).

**Comment by jvitor83** [comment]
Yes. In my case i preffer to not show the cached version of the app, but instead a loading screen while it updates and reload (to avoid a eventual breaking change in API make the cached version unusable or gives error).

**Comment by jvitor83** [comment]
Any news about this? The PR is one year old! 🎂

@gkalpak

**Comment by JeanMeche** [comment]
With https://github.com/angular/angular/pull/63364 we've announced that the service-worker package in now in maintenance mode and will not receive feature PRs and feature requests.

The message is also visible here: https://angular.dev/ecosystem/service-workers
